pandas/core/indexes/multi.py

```diff
@@ -1417,6 +1427,103 @@ def from_product(cls, iterables, sortorder=None, names=None):
+    def from_frame(cls, df, sortorder=None, names=None):
+        """
+        Make a MultiIndex from a DataFrame.
+
+        .. versionadded:: 0.24.0
+
+        Parameters
+        ----------
+        df : pd.DataFrame
+            DataFrame to be converted to MultiIndex.
+        sortorder : int or None
+            Level of sortedness (must be lexicographically sorted by that
+            level).
+        names : list-like, optonal
+            If no names are provided, use the column names, or tuple of column
+            names if the columns is a MultiIndex. If a sequence, overwrite
+            names with the given sequence.
+
+        Returns
+        -------
+        MultiIndex or Index
+            The MultiIndex representation of the given DataFrame.
+
+        Examples
+        --------
+        >>> df = pd.DataFrame([[0, 'happy'], [0, 'jolly'], [1, 'happy'],
+        ...                    [1, 'jolly'], [2, 'joy'], [2, 'joy']],
+        ...                   columns=['number', 'mood'])
+        >>> df
+           number   mood
+        0       0  happy
+        1       0  jolly
+        2       1  happy
+        3       1  jolly
+        4       2    joy
+        5       2    joy
+        >>> pd.MultiIndex.from_frame(df)
+        MultiIndex(levels=[[0, 1, 2], ['happy', 'jolly', 'joy']],
+                   labels=[[0, 0, 1, 1, 2, 2], [0, 1, 0, 1, 2, 2]],
+                   names=['number', 'mood'])
+
+        See Also
+        --------
+        MultiIndex.from_arrays : Convert list of arrays to MultiIndex
+        MultiIndex.from_tuples : Convert list of tuples to MultiIndex
+        MultiIndex.from_product : Make a MultiIndex from cartesian product
+                                  of iterables
+        """
+        from pandas.core.dtypes.generic import ABCDataFrame
+        if not isinstance(df, ABCDataFrame):
+            raise TypeError("Input must be a DataFrame")
+
+        # Get MultiIndex names
+        if names is None:
+            names = df.columns
+        elif is_list_like(names):
+            if len(names) != len(df.columns):
+                    raise ValueError("'names' should have same length as "
+                                     "number of columns in df.")
+        else:
+            raise TypeError("'names' must be a list / sequence of column "
+                            "names.")
+
+        # This way will preserve dtype of columns
+        return cls.from_arrays([df.iloc[:, x] for x in range(len(df.columns))],
+                               sortorder=sortorder,
+                               names=names)
+
+    @classmethod
+    def from_data(cls, data, orient='columns', sortorder=None, names=None):
+        from pandas import DataFrame
+
+        is_df = isinstance(data, DataFrame)
+        try:
+            df = DataFrame(data)
+        except ValueError:
+            raise TypeError("'from_data' input must be valid DataFrame input.")
+        if orient == 'rows':
+            df = df.T
```

Comment:
I would personally prefer a ``rotate=True|False`` parameter. Not a very strong preference, but it is not easy for me to interpret the values of ``orient``.